### PR TITLE
refactor(meta): simplify stream job table/source id assignment

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We fill table_id/source_id in several places, which seems unnecessary. This PR tries to remove unnecessary codes to make it more understandable and maintainable.

We leave it with 2 steps:
1. In `create_job_catalog`, we assign id to `StreamingJob`
2. In `StreamFragmentGraph::new` (`fill_job`), we traverse the nodes, and set the id to corresponding plan nodes. (Note that previously we traverse the nodes multiple times)

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19171?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#19171** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19171?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#19169** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/19169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/meta/src/stream/stream_graph/fragment.rs
Comment (xxchan) [inline src/meta/src/stream/stream_graph/fragment.rs:1]: rename table -> job

File: src/meta/src/stream/stream_graph/fragment.rs [hunk 1]
```diff
@@ -145,27 +145,49 @@ impl BuildingFragment {
...
+                            source_inner.source_id = source.id;
+                        }
+                    }
+                    // For other job types, no need to fill the source id, since it refers to an existing source.
+                    _ => {}
+                }
+            }
+            NodeBody::StreamCdcScan(node) => {
+                if let Some(table_desc) = node.cdc_table_desc.as_mut() {
+                    table_desc.table_id = job_id;
```
Comment (xxchan) [inline src/meta/src/stream/stream_graph/fragment.rs:184]: This replaces the code in `create_streaming_job`

File: src/meta/src/rpc/ddl_controller.rs [hunk 2]
```diff
@@ -918,24 +916,6 @@ impl DdlController {
...
-                // set the inner source id of source node.
-                for fragment in fragment_graph.fragments.values_mut() {
-                    visit_fragment(fragment, |node_body| {
-                        if let NodeBody::Source(source_node) = node_body {
-                            source_node.source_inner.as_mut().unwrap().source_id = src.id;
-                        }
-                    });
-                }
-            }
-            _ => {}
```
Comment (xxchan) [inline src/meta/src/rpc/ddl_controller.rs:936]: do it in `fill_job` instead

File: src/meta/service/src/ddl_service.rs [hunk 3]
```diff
@@ -84,27 +82,28 @@ impl DdlServiceImpl {
...
-        let mut fragment_graph = change.fragment_graph.unwrap();
-        let mut table = change.table.unwrap();
-        if let Some(OptionalAssociatedSourceId::AssociatedSourceId(source_id)) =
-            table.optional_associated_source_id
-        {
-            source.as_mut().unwrap().id = source_id;
-            fill_table_stream_graph_info(&mut source, &mut table, job_type, &mut fragment_graph);
-        }
-        let table_col_index_mapping = change
-            .table_col_index_mapping
```
Comment (xxchan) [inline src/meta/service/src/ddl_service.rs:99]: Note: `extract_replace_table_info` is used in: create/drop sink & table schema change.

File: src/frontend/src/handler/create_table.rs [hunk 4]
```diff
@@ -1441,13 +1442,17 @@ pub async fn generate_stream_graph_for_table(
...
-            .associated_source_id()
-            .map(|source_id| OptionalAssociatedSourceId::AssociatedSourceId(source_id.into())),
         ..table
     };
+    if let Some(source_id) = original_catalog.associated_source_id() {
+        table.optional_associated_source_id = Some(OptionalAssociatedSourceId::AssociatedSourceId(
+            source_id.table_id,
+        ));
+        source.as_mut().unwrap().optional_associated_table_id =
+            Some(OptionalAssociatedTableId::AssociatedTableId(table.id))
```
Comment (xxchan) [inline src/frontend/src/handler/create_table.rs:1454]: It's strange not to set `optional_associated_table_id` here.

---

Note: for replace table, the `Table` and `Source` catalog here were both created again.

File: src/meta/service/src/ddl_service.rs (hunk 3 above)
Comment (xxchan) [reply src/meta/service/src/ddl_service.rs:99]: Some work is done in `fill_job`, and some is done in frontend `generate_stream_graph_for_replace_table`

File: src/meta/src/stream/stream_graph/fragment.rs [hunk 5]
```diff
@@ -145,27 +145,49 @@ impl BuildingFragment {
...
+                dml_node.table_id = job_id;
                 dml_node.table_version_id = job.table_version_id().unwrap();
             }
-            NodeBody::Source(_) => {
-                // Notice: Table job has a dumb Source node, we should be careful that `has_table` should not be overwrite to `false`
-                if !has_table {
-                    has_table = job.is_source_job();
+            NodeBody::Source(source_node) => {
+                match job {
+                    // Note: For table without connector, it has a dumb Source node.
```
Comment (BugenZhao) [inline src/meta/src/stream/stream_graph/fragment.rs:161]:
```suggestion
                    // Note: For table without connector, it has a damn Source node.
```

**Review by BugenZhao — APPROVED** [review]

**Review by yezizp2012 — APPROVED** [review]

**Review by stdrc — COMMENTED** [review]
Appreciate the work done in this PR👍

File: src/meta/src/stream/stream_graph/fragment.rs (hunk 5 above)
Comment (BugenZhao) [reply src/meta/src/stream/stream_graph/fragment.rs:161]:
```suggestion
                    // Note: For table without connector, it has a dummy Source node.
```

**Pull request merged**